### PR TITLE
Re-implement Muuntaja as Protocol

### LINKED ISSUE #59: Re-implement Muuntaja as a Protocol
Currently, it's too easy to modify created Muuntaja instance which has no effect. e.g.

```clj
(-> (m/create) (assoc :default-format "application/edn"))
```

instead of:

```clj
(-> (assoc m/default-options :default-format "application/edn") (m/create))
```

### PR DESCRIPTION
**Pull request by ikitommi**
* Fixes #59 
* Optimized `negotiate-and-format-request` (less hash-lookups)
* Optimized path for UTF-8 with Jsonista
* +20% better e2e throughput with middleware (Jsonista)

screen shot 2018-01-15 at 8 14 44

### CODE REVIEW

File: src/clj/muuntaja/parse.clj [hunk 1]
```diff
@@ -7,16 +7,18 @@
...
     (fn [& args]
       (or (.get cache args)
-          (let [ret (apply f args)]
-            (when ret
-              ;; not synchronized but ok enough
-              (when (> (.size cache) size)
-                (.clear cache))
-              (.putIfAbsent cache args ret))
-            ret)))))
+          (if-not (= args empty)
```
Comment (miikka) [inline src/clj/muuntaja/parse.clj]: Not sure what this is? Always return nil if for `((fast-memoize .. f) nil)`?

File: src/clj/muuntaja/core.clj [hunk 2]
```diff
@@ -8,163 +8,78 @@
...
-                     (and (instance? RuntimeException e)
-                          (= message "EOF while reading"))))
-              (catch Exception _))
-      (throw
-        (ex-info
-          (str "Malformed " format " in " type "")
-          {:type type
-           :format format}
-          e)))))
+(defn consumes [m]
```
Comment (miikka) [inline src/clj/muuntaja/core.clj]: All these small public functions would benefit from docstrings.

**Comment by miikka** [comment]
Since this removes the Muuntaja record, should probably bump the version to `0.5.0` once this is merged.

File: src/clj/muuntaja/parse.clj (hunk 1 above)
Comment (ikitommi) [reply src/clj/muuntaja/parse.clj]: good catch. we need to either guard against input to the memoize or handle the nil-case within it. Without this, the `((fast-memoize .. f) nil)` always misses the cache (`'(nil)` is not found on the cache as it returns `nil` and the `.get` returns `nil`. Not sure why. But the execution cost goes from ~30ns to 3000ns. For cases like "no accept-header set" etc.

File: src/clj/muuntaja/core.clj (hunk 2 above)
Comment (ikitommi) [reply src/clj/muuntaja/core.clj]: indeed.

File: src/clj/muuntaja/parse.clj (hunk 1 above)
Comment (miikka) [reply src/clj/muuntaja/parse.clj]: How about using `.getOrDefault`?

```clojure
(let [cache ...
      sentinel (Object.)]
  (fn [& args]
    (let [cached (.getOrDefault cache args sentinel)]
      (if (identical? sentinel cached)
         (comment compute the value)
         cached))
      
```

**Comment by ikitommi** [comment]
thans for the review, fixed those.

**Pull request merged**